Implement ExecuteDelete

smitpatel opened the pull request:
Part of #795

smitpatel commented:
This covers basic scenario and ability to convert to subquery form. Allowing additional tables is very provider specific. I will add support for it in SqlServer in separate PR. (This PR does improvement for SqlServer when there is only Take specified).

Pending negative cases
- TPT
- TPC
- Entity splitting
- Table splitting
- Non entity

@roji - This is something which can allow you to test against Npgsql.

smitpatel commented on src/EFCore.Relational/Extensions/RelationalQueryableExtensions.cs:
```diff
@@ -224,4 +230,33 @@ public static DbCommand CreateDbCommand(this IQueryable source)
+    #region BulkDelete
+
+    /// <summary>
+    ///     TBD
```
@dotnet/efteam - Can someone suggest docs? :trollface:

roji replied:
Submitted #28523 against this branch with suggestions

smitpatel commented on src/EFCore.Relational/Query/NonQueryExpression.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace Microsoft.EntityFrameworkCore.Query;
+
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
```
Will fill in these later once bulkupdate is implemented.

smitpatel commented on src/EFCore.Relational/Query/QuerySqlGenerator.cs:
```diff
@@ -147,6 +149,40 @@ private static bool IsNonComposedSetOperation(SelectExpression selectExpression)
+        else
+        {
+            // TODO: Exception message
+            throw new InvalidOperationException();
```
@dotnet/efteam - This is where it would throw error if provider said they support more pattern but didn't write custom code for it to generate SQL.

roji replied:
Makes sense, a bit like a debug.assert for providers.

smitpatel commented on src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -946,6 +975,100 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
+                else
+                {
+                    // TODO: Exception message
+                    throw new InvalidOperationException("specific message about keyless");
```
Exception message about keyless entities cannot be converted to subquery form.

roji replied:
Submitted in #28523

smitpatel commented on src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -946,6 +975,100 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
+        }
+
+        // TODO: Throw better exception message indicating that delete requires an entity type
+        return null;
```
Exception message about bulk delete applied on non-entity projection.

roji replied:
Submitted in #28523

smitpatel commented on src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.cs:
```diff
@@ -41,6 +42,94 @@ public partial class RelationalShapedQueryCompilingExpressionVisitor : ShapedQue
+        // Concurrency detector
+        // Exception throwing
+
+        try
```
Does bulk operation requires concurrency detector?
Checking for task cancellation?
What can be other exceptions?

smitpatel replied:
let's discuss in a design meeting.

smitpatel commented on src/EFCore.Relational/Query/SqlExpressions/DeleteExpression.cs:
```diff
@@ -0,0 +1,60 @@
+namespace Microsoft.EntityFrameworkCore.Query.SqlExpressions;
+
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+public sealed class DeleteExpression : Expression, IPrintableExpression
```
Same as NonQueryExpression

smitpatel commented on src/EFCore.SqlServer/Query/Internal/SqlServerQuerySqlGenerator.cs:
```diff
@@ -29,6 +29,50 @@ public class SqlServerQuerySqlGenerator : QuerySqlGenerator
+        }
+        else
+        {
+            throw new NotSupportedException();
```
Same as QuerySqlGenerator

roji replied:
Submitted in #28523

roji commented:
Thanks @smitpatel, will review and implement the PG side against this to validate :+1:

roji reviewed (changes requested):
Look great @smitpatel, definitely simpler/smaller than I feared it may be :cupid:

smitpatel commented on src/EFCore.Relational/Query/SqlExpressions/DeleteExpression.cs:
```diff
@@ -0,0 +1,60 @@
+        => typeof(object);
+
+    /// <inheritdoc />
+    public sealed override ExpressionType NodeType
```
you wanted this in all custom expression at some point lol.

roji replied:
I did??

smitpatel replied:
https://github.com/dotnet/efcore/issues/15662
2nd last point.

roji replied:
Ah sure - my point here was only that it's unnecessary to seal this specific property if the whole class is sealed. In other words, if the CLR type weren't sealed, then yeah, it would make sense to seal the NodeType specifically...

smitpatel commented on src/EFCore.Relational/Query/QuerySqlGenerator.cs:
```diff
@@ -147,6 +149,40 @@ private static bool IsNonComposedSetOperation(SelectExpression selectExpression)
+            && selectExpression.Orderings.Count == 0
+            && selectExpression.GroupBy.Count == 0
+            && selectExpression.Tables.Count == 1
+            && selectExpression.Tables[0] == deleteExpression.Table
```
Set operation are represented as select expression so if table is TableExpression then this doesn't represent set operation by default.

smitpatel commented on src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -156,6 +156,35 @@ when entityQueryRootExpression.GetType() == typeof(EntityQueryRootExpression)
+    /// <inheritdoc />
+    protected override Expression VisitMethodCall(MethodCallExpression methodCallExpression)
+    {
+        NonQueryExpression CheckTranslated(NonQueryExpression? translated)
```
But the base has it at top. We have a pending discussion on convention.

smitpatel commented on src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -946,6 +975,100 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
+    /// <param name="entityShaperExpression">The entity shaper expression on which delete operation is being applied.</param>
+    /// <param name="tableExpression">The table expression from which rows are being deleted.</param>
+    /// <returns> das </returns>
+    protected virtual bool IsValidSelectExpressionForBulkDelete(
```
"Feature" is too subjective. Given there is entity shaper here, if provider wants to throw custom message for keyless when there is offending feature, they can easily do so.

roji replied:
This was more about doing it in relational.. Providers will likely copy whatever we do here in relational (along with error reporting). But not critical.

smitpatel commented on src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.cs:
```diff
@@ -41,6 +42,94 @@ public partial class RelationalShapedQueryCompilingExpressionVisitor : ShapedQue
+        }
+        catch (Exception)
+        {
+            // Throw friendly message
```
Just logging.

smitpatel commented on src/EFCore.SqlServer/Query/Internal/SqlServerQuerySqlGenerator.cs:
```diff
@@ -29,6 +29,50 @@ public class SqlServerQuerySqlGenerator : QuerySqlGenerator
+            Sql.AppendLine($"FROM {Dependencies.SqlGenerationHelper.DelimitIdentifier(deleteExpression.Table.Alias)}");
+
+            Sql.Append("FROM ");
+            Visit(selectExpression.Tables[0]);
```
Yes.

smitpatel replied:
Added support for join.

roji commented:
I thought we used the concurrency detector also because database connections aren't thread-safe... For example, we also use it in ExecuteSqlRaw, which also doesn't have anything to do with the change tracker.

Not sure there's a downside to having it here by default (it can always be disabled for extra perf). /cc @ajcvickers

smitpatel commented on test/EFCore.Relational.Specification.Tests/BulkUpdates/NorthwindBulkUpdatesTestBase.cs:
```diff
@@ -0,0 +1,101 @@
+            ss => ss.Set<OrderDetail>().Where(e => e.OrderID < 10300).Skip(100).Take(100),
+            rowsAffectedCount: 40);
+
+    protected virtual void ClearLog()
```
Added set operations, joins. Will add few more manual joins in inheritance tests.

smitpatel commented:
@roji - Join tests are in for you to light up using syntax.

roji reviewed (approved):
LGTM - seems mergeable (but am happy to continue iterating on this if you want to add more stuff). Synced my PG PR https://github.com/npgsql/efcore.pg/pull/2449 (filtered by last commit) to do join via PG USING.

smitpatel commented on test/EFCore.Relational.Specification.Tests/BulkUpdates/NorthwindBulkUpdatesTestBase.cs:
```diff
@@ -0,0 +1,253 @@
+                            .Select(g => g.First()).Contains(e.Order)),
+            rowsAffectedCount: 40);
+
+    [ConditionalTheory(Skip = "Issue#28525")]
```
It should work hence skipped.

smitpatel commented on src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -946,6 +975,104 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
+    {
+        if (source.ShaperExpression is not EntityShaperExpression entityShaperExpression)
+        {
+            // TODO: Throw better exception message indicating that delete requires an entity type
```
Sorry for renaming the test names.

smitpatel replied:
I will add few more tests around inheritance and query filters (hence added those test bases).

roji replied:
No worries, it's very little work to react, feel free to change anything.

smitpatel commented:
Added test to block all invalid scenarios
- Entity splitting
- Table splitting
- TPT
- TPC non-leaf type
- Non-entity projection
- Keyless entity which requires subquery form

Also added test for query filter, set operations, FromSql, positive cases for TPT/TPC. Using TPT/TPC in predicate while deleting other entity types.

smitpatel commented:
/azp run

roji reviewed (approved):
Here you go @smitpatel. Also synced https://github.com/npgsql/efcore.pg/pull/2449 to the latest state of this.

smitpatel commented on test/EFCore.SqlServer.FunctionalTests/BulkUpdates/NorthwindBulkUpdatesSqlServerTest.cs:
```diff
@@ -0,0 +1,417 @@
+        await base.Delete_Where(async);
+
+        AssertSql(
+            @"DELETE FROM [o]
```
Let me check if it works in SQL server.

roji replied:
Right, I see that this isn't supported with the alias. We could do it without the alias:

```sql
DELETE FROM [Order Details]
WHERE [Order Details].[OrderID] < 10300
```

But this would probably be non-trivial to do in our pipeline. Also it could result in longer SQL if the table name is big and repeated.

smitpatel replied:
Implicit column binding works so table name can be excluded. The implementation in-line with what we had discussed. I think we should leave it customer feedback here because the current way is more maintainable for us.

smitpatel commented on src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -946,6 +970,143 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
+
+        var entityType = entityShaperExpression.EntityType;
+        var mappingStrategy = entityType.GetMappingStrategy();
+        if (mappingStrategy == RelationalAnnotationNames.TptMappingStrategy)
+        {
+            AddTranslationErrorDetails(
+                RelationalStrings.BulkOperationOnTPT(nameof(RelationalQueryableExtensions.BulkDelete), entityType.DisplayName()));
+            return null;
+        }
+
+        if (mappingStrategy == RelationalAnnotationNames.TpcMappingStrategy
+            && entityType.GetDirectlyDerivedTypes().Any())
+        {
+            // We allow TPC is it is leaf type
+            AddTranslationErrorDetails(
+                RelationalStrings.BulkOperationOnTPC(nameof(RelationalQueryableExtensions.BulkDelete), entityType.DisplayName()));
+            return null;
+        }
```
I didn't do this way because there were only 2 labels.

smitpatel commented on src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.cs:
```diff
@@ -41,6 +42,108 @@ public partial class RelationalShapedQueryCompilingExpressionVisitor : ShapedQue
+            }
+            else
+            {
+                relationalQueryContext.QueryLogger.QueryIterationFailed(contextType, exception);
```
Update this to NonQueryFailed.

smitpatel commented on src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.cs:
```diff
@@ -41,6 +42,108 @@ public partial class RelationalShapedQueryCompilingExpressionVisitor : ShapedQue
+        {
+            if (relationalQueryContext.ExceptionDetector.IsCancellation(exception))
+            {
+                relationalQueryContext.QueryLogger.QueryCanceled(contextType);
```
See if we can make it more-generic.

smitpatel commented on src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.cs:
```diff
@@ -41,6 +42,108 @@ public partial class RelationalShapedQueryCompilingExpressionVisitor : ShapedQue
+        }
+    }
+
+
```
empty line

roji commented on test/EFCore.Relational.Specification.Tests/BulkUpdates/NorthwindBulkUpdatesTestBase.cs:
```diff
@@ -0,0 +1,339 @@
+
+    [ConditionalTheory]
+    [MemberData(nameof(IsAsyncData))]
+    public virtual Task Delete_Concat(bool async)
```
What happens for set operations with different source tables? smiling_imp

(trying to think up more negative scenarios for you)

smitpatel replied:
Set operations are always converted to subquery form so regardless of source tables, it just works

roji replied:
But wait, if I do `ctx.Blogs1.Concat(ctx.Blogs2).ExecuteDelete()`, that looks like it should delete all rows from *both* tables, which would mean two SQL statements, no? At least that's what the corresponding query does.

If so, this would possibly be covered by #28520.

smitpatel replied:
Does `ToList` on that query work? If the `Blogs1` & `Blogs2` are 2 different tables then regardless of their CLR types being same (else Concat wouldn't work), their IEntityType must be different. We don't support set operations between 2 different entity types.

roji replied:
OK. It may still be a good idea to add a negative test - if one day we do add support set operations over different entity types, updates would still need to be blocked (unless we implement it via multiple DMLs). But can also forget about it.

smitpatel replied:
> if one day we do add support set operations over different entity types

We cannot support it. Both entity types would have different materialization paths. It is not something we haven't yet added support for, it is something which is not possible to do.

roji replied:
Couldn't we could synthesize a discriminator column if we wanted to - don't we do that for TPT/TPC? In fact a Concat between two entity types seems quite similar to selecting the root of a TPC hierarchy, in both cases two concrete tables are being selected back (and unless I'm mistaken we use a UNION ALL for that).

But anyway...

smitpatel replied:
We decided not to do discriminator thing on #16243

smitpatel replied:
I had more thoughts about this. Have some guidelines about testing, let's discuss in today's design meeting.

smitpatel commented on src/EFCore.Relational/Diagnostics/RelationalEventId.cs:
```diff
@@ -741,6 +742,19 @@ private static EventId MakeQueryId(Id id)
+    ///         This event uses the <see cref="DbContextTypeErrorEventData" /> payload when used with a <see cref="DiagnosticSource" />.
+    ///     </para>
+    /// </remarks>
+    public static readonly EventId BulkOperationFailed = MakeQueryId(Id.BulkOperationFailed);
```
We don't know which operation it is when we are executing it. It is just a non-query expression.

smitpatel replied:
Filed #28568

smitpatel commented on src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:
```diff
@@ -946,6 +970,143 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
+    /// </summary>
+    /// <param name="source">The shaped query on which the operator is applied.</param>
+    /// <returns>The non query after translation.</returns>
+    protected virtual NonQueryExpression? TranslateBulkDelete(ShapedQueryExpression source)
```
This one I will align with the name but we need to have another design meeting about the word "bulk".
While API is more intuitive from name perspective, there is no "execute" operations. The issue title contains "bulk" and for years it has remained that way and no one complained. "bulk operation" as a concept is not a wrong term to use.

jimjimjimhaha replied:
> The issue title contains "bulk" and for years it has remained that way and **no one complained**

Sorry to sneak my way in here, but just wanted to say you probably haven't checked out that thread for a long time. This very word has been complained about over and over again. I thought `Execute` was finalized? It received much more positive feedback.

smitpatel commented on src/EFCore.Relational/Properties/RelationalStrings.resx:
```diff
@@ -127,6 +127,27 @@
   <data name="BadSequenceType" xml:space="preserve">
     <value>Invalid type for sequence. Valid types are long (the default), int, short, byte and decimal.</value>
   </data>
+  <data name="BulkDeleteOnTableSplitting" xml:space="preserve">
```
named variable operation and removed bulk from that. Though there are multiple usage of bulk and no easy way to refer to it. We do need to discuss bulk naming again. (not necessarily change API name but concept is not gone)
